pex/tools/commands/venv.py

```diff
@@ -45,34 +60,38 @@ def _copytree(
             dirs[:] = [d for d in dirs if d not in exclude]
             files[:] = [f for f in files if f not in exclude]
 
-        for d in dirs:
-            try:
-                os.mkdir(os.path.join(dst, os.path.relpath(os.path.join(root, d), src)))
-            except OSError as e:
-                if e.errno != errno.EEXIST:
-                    raise e
-
-        for f in files:
-            src_entry = os.path.join(root, f)
+        for path, is_dir in itertools.chain(
+            zip(dirs, itertools.repeat(True)), zip(files, itertools.repeat(False))
+        ):
+            src_entry = os.path.join(root, path)
             dst_entry = os.path.join(dst, os.path.relpath(src_entry, src))
             yield src_entry, dst_entry
             try:
-                # We only try to link regular files since linking a symlink on Linux can produce
-                # another symlink, which leaves open the possibility the src_entry target could
-                # later go missing leaving the dst_entry dangling.
-                if link and not os.path.islink(src_entry):
-                    try:
-                        os.link(src_entry, dst_entry)
-                        continue
-                    except OSError as e:
-                        if e.errno != errno.EXDEV:
-                            raise e
-                        link = False
-                shutil.copy(src_entry, dst_entry)
+                if symlink:
+                    _relative_symlink(src_entry, dst_entry)
+                elif is_dir:
+                    os.mkdir(dst_entry)
+                else:
+                    # We only try to link regular files since linking a symlink on Linux can produce
+                    # another symlink, which leaves open the possibility the src_entry target could
+                    # later go missing leaving the dst_entry dangling.
+                    if link and not os.path.islink(src_entry):
+                        try:
+                            os.link(src_entry, dst_entry)
+                            continue
+                        except OSError as e:
+                            if e.errno != errno.EXDEV:
+                                raise e
+                            link = False
+                    shutil.copy(src_entry, dst_entry)
             except OSError as e:
                 if e.errno != errno.EEXIST:
                     raise e
 
+        if symlink:
+            # Once we've symlinked the top-level directories and files, we've "copied" everything.
+            return
```

Comment:
What is the purpose of this `if` ?

Reply:
This function is called in one of two modes. If `symlink` is False (the old way), the src is copied to dst, but to save space, hard links are used when possible. When you do this you need to do a full fs walk all the way down src. If `symlink` is `True` then symlinks are used, and to symlink the contents of src you need only symlink each top level src member; then you're done and you can return early without walking further down the src/ tree.

Reply:
That really just re-states the comment in 5x as many words. Maybe drawing things will help - that helps me.

Reply:
Yes, I was aware of that. I had to open up the whole file to spot my "mistake". I failed to notice that this if-block was in the scope of the `for` loop. Sorry for the noise :)

Reply:
Heh, I actually see now, that there's a two nested for loops, but only one of those are shown in the diff...

Reply:
Ah, yeah. I pretty much <3 GitHub Diffs as much as I <3 macOS.

Reply:
This bit (only copying/linking the root directories, rather than every inner node) is one reason why I think that this improves space usage: you're right that hardlinking did not create new inodes per file, but it did create new inodes per directory, so the directory listings ended up fully duplicated.

Either way: nice!